ecmascript/dep-graph/src/lib.rs

```diff
@@ -0,0 +1,304 @@
+use swc_common::{
+    comments::{Comment, SingleThreadedComments},
+    Loc, DUMMY_SP,
+};
+pub use swc_common::{SourceMap, Span};
+use swc_ecma_ast as ast;
+use swc_ecma_visit::{self, Node, Visit, VisitWith};
+
+pub fn analyze_dependencies(
+    module: &ast::Module,
+    source_map: &SourceMap,
+    comments: &SingleThreadedComments,
+) -> Vec<DependencyDescriptor> {
+    let mut v = DependencyCollector {
+        comments,
+        source_map,
+        items: vec![],
+    };
+    module.visit_with(&ast::Invalid { span: DUMMY_SP }, &mut v);
+    v.items
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum DependencyKind {
+    Import,
+    ImportType,
+    Export,
+    ExportType,
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct DependencyDescriptor {
+    pub kind: DependencyKind,
+    /// A flag indicating if the import is dynamic or not.
+    pub is_dynamic: bool,
+    /// Any leading comments associated with the dependency.  This is used for
+    /// further processing of supported pragma that impact the dependency.
+    pub leading_comments: Vec<Comment>,
+    /// The location of the import/export statement.
+    pub col: usize,
+    pub line: usize,
+    /// The text specifier associated with the import/export statement.
+    pub specifier: String,
+}
+
+struct DependencyCollector<'a> {
+    comments: &'a SingleThreadedComments,
```

Comment:
SingleThreadedComments seems bit opinionated. How about using `&'a dyn Comments?`

Reply:
Sure!

Reply:
Minor inconvenience is that `Comments` has no `with_leading` method - I can work around it by taking a comment, cloning it, and putting it back into the map. Does that work for you?

Reply:
It works for me but I think using `SingleThreadedComments` is better than it.